src/lib-zephyr/cpu.c

```diff
@@ -0,0 +1,192 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Tomasz Leman <tomasz.m.leman@intel.com>
+
+/**
+ * \file
+ * \brief Zephyr RTOS CPU implementation file
+ * \authors Tomasz Leman <tomasz.m.leman@intel.com>
+ */
+
+#include <sof/init.h>
+#include <sof/lib/cpu.h>
+#include <sof/lib/pm_runtime.h>
+#include <rtos/wait.h>
+#include <sof/trace/trace.h>
+
+/* Zephyr includes */
+#include <soc.h>
+#include <version.h>
+#include <zephyr/kernel.h>
+#include <zephyr/pm/pm.h>
+
+LOG_MODULE_DECLARE(zephyr, CONFIG_SOF_LOG_LEVEL);
+
+extern struct tr_ctx zephyr_tr;
+
+#if CONFIG_MULTICORE && CONFIG_SMP
+
+extern K_KERNEL_STACK_ARRAY_DEFINE(z_interrupt_stacks, CONFIG_MP_NUM_CPUS,
+				   CONFIG_ISR_STACK_SIZE);
+
+static atomic_t start_flag;
+static atomic_t ready_flag;
+
+#ifndef CONFIG_PM
+static int w_core_enable_mask = 0x1; /*Core 0 is always active*/
+#endif
```

Comment:
I think this is failing the TGL_IPC4 CI test. We don't have Zephyr CONFIG_PM enabled for TGL, but TGL still assumes that cores can be powered up and down (i.e. mock to work).

Reply:
@tmleman maybe surround this with "ifdef CONFIG_ZEPHYR_NATIVE_DRIVERS" . On MTL builds this is defined, and in TGL this is only defined when we move to Zephyr native drivers. This will add some additional work for @juimonen and me, but this is ok and should get the CI to be green with this PR.

Reply:
@kv2019i sorry, but i don't get it. TGL don't have enabled CONFIG_PM and want to use this local static value. Am I right?

Reply:
@tmleman Ack, sorry, that was not correct what I write above. Let me try to reproduce the failure locally on a TGL hw and see what happen. I'll report back here.

Reply:
Ok, this is more complicated. In old pre-Zephyr SOF, implementation of cpu_enable_core() has been done via pm_runtime interface by doing:
pm_runtime_get(PM_RUNTIME_DSP, PWRD_BY_TPLG | id);

With current Zephyr TGL, we still go through the runtime-pm layer. Your patch removes the call to pm_runtime_get() on TGL (when powering up a core) and this causes the bookkeeping of which cores are active to go out of sync and thus the errors in CI.

Let me try to see if we can fix this somehow without more major reworks. There are still multiple calls to pm_runtime() from SOF app code, so we need to figure out a transition to Zephyr native APIs. But in short term we can proceed if we can keep TGL functional at its current level.

Reply:
My attempt at fixing this, pushed at https://github.com/thesofproject/sof/pull/6483 . If test s fine, @tmleman please include these changes to your PR and resubmit.

Also, please take a look at the pm-runtime connection. Now pm-runtime (in SOF) e.g. keeps track of whether a core was powered-up because it was needed to enable a clock, or whether it was needed as it is used in topology. I think Zephyr will natively handle the former case, so we are left with only tplg users (I think). But worth double-checking that we don't have any SOF side application logic in pm-runtime, that we'd still need to do in SOF, on top of Zephyr CONFIG_PM.

Reply:
More tests pass, but there's still failure with #6483 in multicore Zephyr IPC3 tests. :( Errors in FW log:
```
[    60607025.925026] (       38124.164062) c0 ipc                  src/ipc/ipc3/handler.c:1606 INFO ipc: new cmd 0x40070000
[    60607048.216692] (          22.291666) c0 ipc                  src/ipc/ipc3/handler.c:717  INFO ipc: pm core mask 0x5 -> enable
[    60607609.831253] (         561.614563) c0 ipc                  src/ipc/ipc3/handler.c:1606 INFO ipc: new cmd 0x30100000
[    60607633.060419] (          23.229166) c0 ipc                    src/ipc/ipc-common.c:51   ERROR ipc_process_on_core(): core #2 is disabled
```

Reply:
> Ok, this is more complicated. [...] this causes the bookkeeping of which cores are active to go out of sync and thus the errors in CI.

Do you mean these errors? https://sof-ci.01.org/sofpr/PR6041/build2053/devicetest/ Always copy the URLs to failures because they're hard/impossible to find after a force-push

The problem looks complicated but the reproduction should not be. @tmleman https://github.com/thesofproject/sof-test should be enough to reproduce, if not please contact me. No validation "secrets".

If you don't have any TGL or ADL device you can borrow one, please contact me about that too.